[Vagrantfile] Npm install support for Windows

ryansnowboarder opened the pull request:
"vagrant up" previously failed at "npm install" because of attempts at symbolic linking on Windows, which won't work at all. Solved problem by adding "--no-bin-links" flag to the command to stop symbolic linking for the files being installed. I assumed this change wouldn't cause a problem on Linux host os, but if it does, I can create a patch for only adding the flag if the host os is Windows.

smarx commented:
*Automated message from Dropbox CLA bot*

@ryansnowboarder, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented:
Thanks for working on this @ryansnowboarder!  I have a few comemnts:
* You should rebase this branch and squash the commits together -- the history of the typo you made in the first version isn't useful :).  
* I'd prefer to do this only on windows.

Also it's worth investigating if there's some other way we should be fixing this category of problem, since `npm` is probably not the only place we will ever use code that tries to create a symbolic link in the Zulip development environment; e.g. this article seems to suggest changing Virtualbox settings as being relevant (though the top 2 answers disagree about what to do)...

http://stackoverflow.com/questions/24200333/symbolic-links-and-synced-folders-in-vagrant

ryansnowboarder commented:
Ok, yeah I agree. I'll look into the SO methods.

On Sun, Mar 27, 2016 at 3:21 PM, Tim Abbott <notifications@github.com>
wrote:

> Thanks for working on this @ryansnowboarder
> <https://github.com/ryansnowboarder>! I have a few comemnts:
>
>    - You should rebase this branch and squash the commits together -- the
>    history of the typo you made in the first version isn't useful :).
>    - I'd prefer to do this only on windows.
>
> Also it's worth investigating if there's some other way we should be
> fixing this category of problem, since npm is probably not the only place
> we will ever use code that tries to create a symbolic link in the Zulip
> development environment; e.g. this article seems to suggest changing
> Virtualbox settings as being relevant (though the top 2 answers disagree
> about what to do)...
>
>
> http://stackoverflow.com/questions/24200333/symbolic-links-and-synced-folders-in-vagrant
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/zulip/zulip/pull/593#issuecomment-202127402>
>

timabbott commented:
OK, cool.  I think if those aren't promising or it'll take a long time to investigate, I'd certainly merge something to add that flag when running inside a windows host, since that will make provisioning work on windows.

ryansnowboarder commented:
I've put the task on a pretty high priority for this weekend and I'll let
you know the results by the end of Sunday.

On Sat, Apr 2, 2016 at 1:23 AM, Tim Abbott <notifications@github.com> wrote:

> OK, cool. I think if those aren't promising or it'll take a long time to
> investigate, I'd certainly merge something to add that flag when running
> inside a windows host, since that will make provisioning work on windows.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/zulip/zulip/pull/593#issuecomment-204649164>
>

ryansnowboarder commented:
@timabbott I've done the following and hopefully this pull request is good to go for a merge:

1) Fixed up README.dev.md as requested.
2) Looked into the SO methods:

The symbolic linking is an issue specific to using Vagrant with Virtualbox, which does not necessarily only happen on Windows. Thus, patched vagrantfile with the rsync fix for Vagrant in Virtualbox found in answer #2 in the SO topic (http://stackoverflow.com/questions/24200333/symbolic-links-and-synced-folders-in-vagrant). Answer #1 appears to be outdated since Vagrant 1.1 and did not work in  Vagrant 1.8.  

3) Some newer versions of Vagrant have a known bug with rsync, so I linked to a solution in README.dev.md 

4) Rebased my branch and squashed many commits.

timabbott commented on README.dev.md:
```diff
@@ -4,66 +4,90 @@ Installing the Zulip Development environment
+
+Windows:
+--------
+`git clone https://github.com/zulip/zulip.git  -c core.autocrlf=false` 
```
should get rid of the double-space here

ryansnowboarder replied:
Working on it now.

timabbott commented on README.dev.md:
```diff
@@ -4,66 +4,90 @@ Installing the Zulip Development environment
-  version of Vagrant][vagrant-dl], and then do the following:
+Ubuntu 14.04
+------------
+*  [download a newer  version of Vagrant][vagrant-dl], and then do the following:
```
should fix the spacing here

ryansnowboarder replied:
Working on it now.

timabbott commented on Vagrantfile:
```diff
@@ -31,6 +31,7 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
   end
 
   config.vm.provider "virtualbox" do |vb, override|
+    config.vm.synced_folder ".", "/srv/zulip", type: "rsync", rsync__args: ["--verbose", "--archive", "--delete", "-z"]
```
Can you add a comment here explaining and/or linking to the article, docs, or whatever why this is correct?

ryansnowboarder replied:
It's the second answer in the SO topic you posted earlier (http://stackoverflow.com/questions/24200333/symbolic-links-and-synced-folders-in-vagrant). Per the Vagrant docs, "Vagrant can use rsync as a mechanism to sync a folder to the guest machine... The rsync synced folder type accepts the following options:

rsync__args (array of strings) - A list of arguments to supply to rsync. By default this is ["--verbose", "--archive", "--delete", "-z", "--copy-links"]."

--copy-links is causing the linking problem, so we configure the synced_folder without it

timabbott commented:
Awesome, thanks for working on this @ryansnowboarder !  I posted a few comments.

timabbott commented on README.dev.md:
```diff
@@ -4,66 +4,90 @@ Installing the Zulip Development environment
+On Windows,  run `npm install --no-bin-links`, and scroll to the
+"All Systems" instructions located way far below. Start from after
+`npm install`. The problem is symbolic linking is being attempted 
+on Windows and crashes `vagrant up`. 
```
I'd prefer to not need to tell users to run some fixup command if we can avoid it; is it possible for us to pass as an option to provision.py whether we're running on Windows, and if so, pass that argument to `npm install` automatically?

ryansnowboarder replied:
I've fixed it so it shouldn't fail on Windows with the patch to Vagrantfile. I should have squashed this commit.

ryansnowboarder commented:
@timabbott Should be good to go!

timabbott commented:
A few quick requests:
* Can you squash the commits fixing spacing/colon errors into the commits that introduced the issues, so each individual commit is clean?
* Can you fix the spelling of "kernel"?

ryansnowboarder commented:
Yes!

ryansnowboarder commented:
@timabbott  done!

timabbott commented:
Closing this since @christi3k has since added working windows instructions.
